src/QAT/qat/purr/backends/verification.py

```diff
@@ -0,0 +1,67 @@
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright (c) 2023 Oxford Quantum Circuits Ltd
+
+from abc import ABC
+
+from qat.purr.backends.live import LiveHardwareModel
+from qat.purr.compiler.emitter import QatFile
+from qat.purr.compiler.execution import QuantumExecutionEngine
+
+
+class QPUVersion:
+    def __init__(self, make: str, version: str = None):
+        self.make = make
+        self.version = version
+
+    def __repr__(self):
+        return f"{self.make}-{self.version or 'latest'}"
+
+
+class Lucy:
+    # TODO: Automate this from class name, not trivially doable due to the class still being
+    #  initialized at this point. Don't want to do deferred exe for such a simple case.
+    _name = "Lucy"
+
+    Latest = QPUVersion(_name)
+
+
+class VerificationModel(LiveHardwareModel):
+    def __init__(self, qpu_version, verification_engine_type: type):
+        super().__init__(None, [verification_engine_type], None)
+        self.version = qpu_version
+
+
+def get_verification_model(qpu_type: QPUVersion):
+    """
+    Get verification model for a particular QPU make and model. Each make has its own class, which has a field
+    that is each individual version available for verification.
+
+    For example, if you wanted to verify our Lucy machine, that'd be done with:
+    ``
+    get_verification_model(Lucy.Latest)
+    ``
+
+    Or with a specific version:
+    ``
+    get_verification_model(Lucy.XY)
+    ``
+    """
+    if not isinstance(qpu_type, QPUVersion):
+        raise ValueError(f"{qpu_type} is not a QPU version, can't find verification engine.")
+
+    if qpu_type.make == Lucy.__name__:
+        # TODO: Should have an apply_setup_to_hardware in live which people can use to build our own architecture.
+        model = VerificationModel(qpu_type, LucyVerificationEngine)
+        raise NotImplementedError("No lucy-specific model created yet.")
+
+    return None
+
+
+class VerificationEngine(QuantumExecutionEngine, ABC):
```

Comment:
Think it's worth adding an abc verify_instructions on this and the model? I like the override api, but also having an instructions verifier wrapped in a results object would be clean for anyone looking to submit instructions over the cloud

Reply:
You mean a free-standing method that just takes in instructions? I mean that'd be fine, but people will need to know that you _have_ to have already passed the verifying instructions through the full pipeline before calling it. You can't just parse them and verify.

Also they won't be in results - it'll just have a composite exception thrown with all the issues with it, unless you meant something else?

Reply:
I've added a free-standing method for people to use, but these verifiers will never return a result - they will just exception with everything wrong in the messages. It's a verifier for correctness, not inefficiencies.